src/components/TrackListItem.js

```diff
@@ -0,0 +1,112 @@
+import PropTypes from "prop-types";
+import {
+    Container,
+    Row,
+    Col
+  } from "reactstrap";
+import {TrackFilePicker} from './TrackFilePicker';
+import {TrackTypeDropdown} from './TrackTypeDropdown';
+import {TrackDeleteButton} from './TrackDeleteButton';
+import {TrackSettingsButton} from './TrackSettingsButton';
+import React, { useEffect, useReducer, useRef } from 'react';
+
+
+export const TrackListItem = ({
+    trackProps,
+    // trackProps expects an object with
+      // trackType: string
+      // trackFile: file object / undefined
+      // availableTracks: array of tracks(see types.ts)
+      // trackColorSettings: object(aka. colorScheme)
+      // availableColors: array of ColorPalletes
+    onChange, // expects a new trackProps object
+    onDelete,
+  }) => {
+    const [myTrackProps, dispatch] = useReducer(reducer, trackProps);
+    const _onChange = useRef(onChange);
+
+    // https://overreacted.io/a-complete-guide-to-useeffect/
+    function reducer(state, action) {
+      let newState = {...state};
+      switch (action.type){
+        case "setType":
+          newState["trackFile"] = undefined;
+          newState["trackType"] = action.payload;
+          break;
+        case "setFile":
+          newState["trackFile"] = action.payload;
+          break;
+        case "setColorSettings":
+          newState["trackColorSettings"] = action.payload;
+          break;
+        default:
+          throw new Error();
+      }
+      return newState;
+    }
+  
+    const trackTypeOnChange = async(newTrackType) => {
+      // wait until file change to call onChange
+      dispatch({type: "setType", payload: newTrackType});
+    }
+
+    const trackFileOnChange = async(newFile) => {
+      dispatch({type: "setFile", payload: newFile});
+    }
+
+    const trackSettingsOnChange = async(key, value) => {
+      let newTrackColorSettings = {...trackProps["trackColorSettings"]};
+      newTrackColorSettings[key] = value;
+      dispatch({type: "setColorSettings", payload: newTrackColorSettings});
+    }
+
+    // https://stackoverflow.com/questions/55840294/how-to-fix-missing-dependency-warning-when-using-useeffect-react-hook
+    // useEffect hook to tell react to call onchange after state changes
+    useEffect(() => {
+      if (myTrackProps.trackFile !== undefined){
+        _onChange.current(myTrackProps);
+      }
```

Comment:
Isn't this going to call onChange on the first render, as long as `trackFile` is set in the props?

(And also if the user changes the track type, then changes it back and re-picks the initial file? Though that might be OK).

I feel like we should only call onChange when there's an actual change. So maybe we want to skip the call if myTrackProps is deep-equal to trackProps?